### PR TITLE
fix(agent): recover finalized private catalog rows from VM

### PR DESCRIPTION
**Pull request by branarakic-agent**
## Summary

- allow retained finalized private inventory rows to rebuild from their exact VM graph after the SWM workspace head is retired
- require the VM fallback to carry a confirmed metadata envelope matching the requested UAL, assertion version, batch, public/private counts, Merkle roots, and canonical graph
- read the exact VM projection through the canonical paged storage helper
- preserve the existing fail-closed workspace-head requirement for public lanes
- funnel inventory-row recovery and confirmed private repair through one internal discriminated resolver, with the existing exports kept as thin adapters
- verify the canonical projection digest against signed inventory rows

## Release certification failure

The local RFC-64 10/11 blackbox runs reached finalized VM parity but stalled at 8/20 SWM catalog recovery. Finalized private placements intentionally remain in the tier-neutral author inventory after their byte-identical SWM twins are retired. Cold reconciliation incorrectly required those retained rows to still have an SWM workspace head and rejected them with `durable RFC-64 workspace head differs`.

## Validation

- `pnpm --filter @origintrail-official/dkg-agent test:unit`
  - 241 test files passed
  - 3,496 tests passed; 5 skipped; 0 failed
- `pnpm --filter @origintrail-official/dkg-agent exec tsc --noEmit`
- `pnpm --filter @origintrail-official/dkg-agent build`
- `pnpm lint`
- focused resolver, inventory reconciler, and native-wiring integration suites
- `git diff --check`

## Regression coverage

The resolver tests cover private VM fallback, public-lane rejection, signed inventory digest mismatch, same-count VM corruption, and a missing VM graph. They also prove that a stale confirmed v1 VM graph cannot satisfy a headless v2 private update with identical public bytes but a different private commitment, while exact confirmed-v2 metadata succeeds. The native-wiring integration fixture now mirrors production finalized VM state by persisting its confirmed metadata envelope.

### CODE REVIEW

File: packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts [hunk 1]
```diff
@@ -71,6 +71,23 @@ export async function resolveRfc64InventoryWorkspaceCatalogAssetV1(
     kaUal: params.row.kaUal,
   });
   throwIfAbortedV1(params.signal);
+  // Finalized private placements remain in the tier-neutral author inventory
+  // after their byte-identical SWM twin is retired. Rebuild those rows from
+  // the exact VM graph; public lanes must continue to require a workspace head.
+  if (head === undefined && params.laneKind === 'private') {
```
Comment (otReviewAgent) [inline packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: 🟡 Issue: Keep one projection-validation pipeline instead of adding a private-lane early return

**What's wrong**
The fallback is bolted onto the inventory resolver as a second terminal path. That duplicates the common validation/materialization tail and separates VM-backed inventory rows from workspace-backed rows even though both must satisfy the same inventory contract. This makes the function more branch-sensitive and invites future validation drift.

**Example**
Any future invariant applied after projection encoding must now be added to both the private VM branch and the workspace branch; the projection-digest check and `catalogAssetV1` call are already duplicated this way.

**Suggested direction**
Make the new branch select the projection source rather than terminate the resolver. Assign `projectionBytes` from either the validated workspace snapshot or the allowed finalized-VM fallback, then run the row digest check and construct the catalog asset once.

**For Agents**
Refactor `resolveRfc64InventoryWorkspaceCatalogAssetV1` so the head/fallback branch only selects `projectionBytes`, followed by one inventory-digest check and one `catalogAssetV1` return. Consider sharing the workspace-or-finalized-VM selection with `resolveRfc64ConfirmedVmRepairCatalogAssetV1` only if that can remain direct. Preserve the rule that public lanes require a workspace head while private lanes may use finalized VM, and verify both sources pass through the same post-resolution validation.

Comment (branarakic-agent) [reply packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: Addressed in b94d4b1f3. The head/fallback branch now only selects projectionBytes. Both workspace-backed and finalized-private-VM projections pass through one shared signed inventory digest check and one catalogAssetV1 construction. Public lanes still fail closed when the workspace head is absent. Validation: focused resolver and inventory reconciler suites 8/8, TypeScript noEmit, agent build including type/package-root tests, direct oxlint, and git diff --check all pass.

Comment (otReviewAgent) [reply packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: 🟡 Issue: Collapse the two parallel asset-resolution workflows

**What's wrong**
This change adds the VM fallback by expanding the inventory resolver into a second copy of the existing repair resolver. The duplication splits ownership of projection-source selection and workspace validation across two functions, increasing the concepts and branches maintainers must keep synchronized.

**Example**
A future change to workspace snapshot loading, abort handling, or seal validation must be applied consistently to both branches. The two paths can silently drift because they independently own the same projection-selection workflow.

**Suggested direction**
Use one internal resolver that performs seal/head resolution, selects the VM or workspace source once, validates and encodes once, then applies source-specific constraints such as share-operation and projection-digest checks. The exported functions can disappear or become minimal typed adapters.

**For Agents**
Refactor packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts around one durable catalog asset resolver with a discriminated input for inventory-row versus confirmed-VM repair constraints. Preserve public-lane missing-head rejection, private VM fallback, workspace-head validation, and inventory projection-digest validation. Existing resolver tests should prove both source variants retain their behavior.

File: packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts [hunk 2]
```diff
@@ -71,28 +71,45 @@ export async function resolveRfc64InventoryWorkspaceCatalogAssetV1(
...
+
+  let projectionBytes: Uint8Array;
+  // Finalized private placements remain in the tier-neutral author inventory
+  // after their byte-identical SWM twin is retired. Rebuild those rows from
+  // the exact VM graph; public lanes must continue to require a workspace head.
+  if (head === undefined) {
+    if (params.laneKind !== 'private') {
+      throw new Error(`durable RFC-64 workspace head differs for ${params.row.kaUal}`);
+    }
+    projectionBytes = await resolveFinalizedVmProjectionBytesV1(
```
Comment (otReviewAgent) [inline packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: 🔴 Bug: The VM fallback does not prove that this assertion version is confirmed

**What's wrong**
A missing private workspace head is treated as proof that the inventory row is a finalized VM placement. However, the VM graph name is shared across assertion versions, and the helper validates only its public bytes against the requested row’s seal. A stale VM graph can therefore satisfy a newer private-only assertion whose public projection is unchanged, allowing catalog reconciliation to sign a row from the wrong lifecycle state.

**Example**
Suppose confirmed VM v1 contains public quads P and private root A. A pending v2 private-only update retains P but has private root B. If the v2 SWM head disappears, the stable VM graph still contains P while its metadata identifies v1/A. This fallback combines P with the v2 seal’s B, successfully validates the v2 root and projection digest, and returns v2 as a finalized VM-backed asset even though only v1 is confirmed. It should reject because the VM control metadata does not prove v2/B.

**Suggested direction**
Resolve and validate the confirmed VM metadata envelope—similar to `resolveConfirmedGraphScopedVm`—before using the graph bytes as a fallback. Public-byte and seal-root equality alone cannot establish the VM assertion version or private commitment.

**For Agents**
In `resolveFinalizedVmProjectionBytesV1`, validate the current confirmed VM metadata before accepting its graph: require the requested UAL, assertion version, Merkle root, public/private counts, private root, and canonical graph URI. Preserve the existing workspace-head behavior and public-lane rejection. Add a regression where confirmed v1 and a headless v2 private-only update share public quads but have different private roots; v2 must fail, while exact confirmed-v2 metadata must pass.

File: packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts [hunk 3]
```diff
@@ -162,6 +167,29 @@ export async function resolveRfc64ConfirmedVmRepairCatalogAssetV1(
...
+  identity: Readonly<Rfc64DurableCatalogAssetIdentityV1>,
+  seal: Readonly<CanonicalGraphScopedAuthorSealV1>,
+  source: string,
+): Promise<Uint8Array> {
+  const vmGraph = knowledgeAssetLayerGraphUri(
+    params.contextGraphId,
+    MemoryLayer.VerifiableMemory,
+    createGraphKnowledgeAssetScope(identity.kaUal, identity.assertionVersion),
+  );
+  const result = await params.store.query(
```
Comment (otReviewAgent) [inline packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: 🟡 Issue: Reuse the canonical exact-graph reader

**What's wrong**
The extracted helper promotes an ad-hoc SPARQL graph read into shared catalog infrastructure even though the storage package already owns this abstraction. That duplicates low-level query/result/count handling in the wrong layer and makes the resolver bypass future improvements to the canonical utility.

**Example**
resolveFinalizedVmProjectionBytesV1 currently recreates only a subset of the storage helper's exact-read policy. Any improvements to paging, response bounds, duplicate detection, result validation, or adapter compatibility in readExactGraphPaged will not automatically apply here.

**Suggested direction**
Delegate graph retrieval and structural validation to readExactGraphPaged with expectedQuadCount and outputGraph: ''. Keep this module responsible only for locating the VM graph and validating its projection against the author seal.

**For Agents**
Replace the raw query in resolveFinalizedVmProjectionBytesV1 with the canonical storage exact-graph reader. Pass the seal's public triple count, graphless output if required by projection encoding, and the existing source/signal query options. Preserve the resolver's domain-specific error and seal-root validation; verify exact, missing, and divergent VM projections still produce the intended outcomes.

File: packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts (hunk 1 above)
Comment (branarakic) [reply packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: Addressed in 541ed0d34. Both exported workflows are now thin adapters over one internal discriminated resolver. Seal/head resolution, projection selection, seal/root validation, canonical encoding, and catalog-asset construction are shared; only inventory-specific lane/share/digest checks and source labels differ.

File: packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts (hunk 2 above)
Comment (branarakic) [reply packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: Addressed in 541ed0d34. The headless private fallback now requires an exact confirmed VM metadata envelope matching the requested UAL, assertion version, batch/reserved KA id, public/private counts, public/private Merkle commitments, and canonical graph URI before reading any bytes. The new regression rejects confirmed v1 versus headless v2 with identical public quads and different private roots; the exact confirmed-v2 companion passes.

File: packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts (hunk 3 above)
Comment (branarakic) [reply packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: Addressed in 541ed0d34. The ad-hoc CONSTRUCT query is replaced with readExactGraphPaged using expectedQuadCount, graphless output, and the existing source/signal. Exact-read integrity failures retain the resolver's domain-specific durable finalized VM projection error.

File: packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts (hunk 2 above)
Comment (otReviewAgent) [reply packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: 🔴 Bug: The VM fallback rejects valid projections above 100,000 triples

**What's wrong**
The new exact reader silently inherits a generic 100,000-quad cap. That cap is narrower than the RFC-64 projection contract and the finalized VM materializer, so a valid confirmed private asset can become permanently unrecoverable by catalog reconciliation after its workspace head is retired.

**Example**
A confirmed private VM asset with 100,001 compact triples can be successfully materialized and fit within the catalog bundle limit. With no workspace head, this resolver now throws QUAD_COUNT_LIMIT_EXCEEDED instead of rebuilding its catalog asset.

**Suggested direction**
Supply an explicit maxQuadCount based on the catalog projection's established resource ceiling, and an aligned byte ceiling, rather than inheriting the storage helper's narrower generic default.

**For Agents**
Update resolveFinalizedVmProjectionQuadsV1 to pass explicit graph-read limits aligned with the RFC-64 catalog projection limits while preserving exact count, byte, Merkle-root, metadata, and abort validation. Add a headless confirmed-VM regression case above 100,000 triples that resolves successfully.

File: packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts (hunk 3 above)
Comment (otReviewAgent) [reply packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: 🟡 Issue: Reuse the canonical confirmed-VM resolver instead of creating a second recognition pipeline

**What's wrong**
This change reimplements a responsibility that the agent already deliberately centralizes. The local graph construction, metadata-envelope checks, byte equality helpers, bounded graph read, error mapping, and subsequent Merkle validation create a parallel definition of “exact confirmed graph-scoped VM.” That increases concepts and makes the two callers liable to drift as storage metadata evolves.

**Example**
If resolveConfirmedGraphScopedVm later changes its metadata acceptance or ExactGraphReadError handling, finalization and materialization will adopt the change while this catalog fallback continues using its private rules for the same confirmed VM graph.

**Suggested direction**
Keep RFC-64-specific inventory validation in this module, but delegate confirmed VM metadata, identity, and exact-content recognition to resolveConfirmedGraphScopedVm. If its current contract lacks cancellation or expected count fields, expand that shared boundary rather than duplicating its implementation.

**For Agents**
Replace resolveFinalizedVmProjectionQuadsV1's manual recognition pipeline with resolveConfirmedGraphScopedVm. Extend that canonical resolver to accept the abort signal, query source, and any additional seal expectations needed here. Preserve the existing RFC-64 error messages, exact public/private counts, private root, no-subgraph requirement, and inventory projection-digest check. Prove with the resolver and materializer tests that identical VM fixtures receive identical recognition decisions.

File: packages/agent/test/rfc64-swm-catalog-durable-asset-resolver-v1.test.ts [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+      ...PROJECTION_QUADS[0]!,
+      object: '"Mallory"',
+    }]);
+
+    await expect(resolve('private')).rejects.toThrow(
+      `durable finalized VM projection differs for ${seal.kaUal}`,
+    );
+  });
+
+  it('rejects a missing finalized VM projection', async () => {
```
Comment (otReviewAgent) [inline packages/agent/test/rfc64-swm-catalog-durable-asset-resolver-v1.test.ts:99]: 🔴 Bug: The missing-projection test exits at the metadata gate

**What's wrong**
This negative test gives false confidence about the new exact finalized-graph validation. It succeeds because mandatory metadata is missing, not because the projection is missing, so the behavior named by the test remains unverified.

**Example**
Seed valid metadata but no projection, for example with `await seedVmProjection(store, seal, [])`, then assert that private resolution rejects with the finalized-VM projection error.

**Suggested direction**
Create a metadata-only fixture, or seed normally and remove only the VM graph quads, so the test reaches readExactGraphPaged and exercises the newly added integrity-error handling.

**For Agents**
Update the missing-projection case in this test file so confirmed metadata matches the seal while the VM graph is empty or absent. The test should prove that the exact graph read rejects the count mismatch and that the resolver preserves its fail-closed error contract.

Comment (branarakic) [reply packages/agent/test/rfc64-swm-catalog-durable-asset-resolver-v1.test.ts:99]: Addressed in d769d4f47. The missing-projection case now seeds a matching confirmed metadata envelope with zero VM data quads, so resolution reaches readExactGraphPaged, observes the expected-count mismatch, and preserves the domain-specific fail-closed error. I also found that this new resolver test file was absent from the RFC-64 unit manifest; it is now registered so CI actually executes all seven cases. Focused resolver + reconciler validation is 10/10 green, with typecheck, lint, and diff checks green.

File: packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts (hunk 1 above)
Comment (otReviewAgent) [reply packages/agent/src/rfc64/swm-catalog-durable-asset-resolver-v1.ts]: 🟡 Issue: The resolution union moves complexity instead of removing it

**What's wrong**
The new union and pass-through wrappers reduce textual duplication at the cost of interleaving two distinct policies. Their differences remain, but are now scattered across five discriminator checks inside one busy control flow.

**Example**
A confirmed-VM repair is represented indirectly as `laneKind = 'private'`, while inventory-only behavior is distributed across `inventoryRowDiffers`, the query-source selection, and the final projection-digest branch. Understanding either public operation therefore requires following the other operation's branches too.

**Suggested direction**
Dispatch once at the public entry points and share a small projection-validation/asset-construction tail, rather than routing both contracts through a discriminator that is rechecked throughout the shared function.

**For Agents**
Restructure the two exported resolvers so each owns its mode-specific workspace/fallback policy. Retain shared helpers only for genuinely identical work such as strict-seal resolution, verified projection encoding, and canonical confirmed-VM loading. Preserve both public signatures and all current validation behavior.
